File: 2022/day14/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type Point struct {
	x, y int
}

func parsePaths(text string) map[Point]bool {
	rocks := make(map[Point]bool)
	lines := strings.Split(strings.TrimSpace(text), "\n")

	for _, line := range lines {
		parts := strings.Split(line, " -> ")
		for i := 0; i < len(parts)-1; i++ {
			p1 := parsePoint(parts[i])
			p2 := parsePoint(parts[i+1])

			// Draw line from p1 to p2
			if p1.x == p2.x {
				// vertical line
				minY, maxY := min(p1.y, p2.y), max(p1.y, p2.y)
				for y := minY; y <= maxY; y++ {
					rocks[Point{p1.x, y}] = true
				}
			} else {
				// horizontal line
				minX, maxX := min(p1.x, p2.x), max(p1.x, p2.x)
				for x := minX; x <= maxX; x++ {
					rocks[Point{x, p1.y}] = true
				}
			}
		}
	}
	return rocks
}

func parsePoint(s string) Point {
	parts := strings.Split(s, ",")
	x, _ := strconv.Atoi(parts[0])
	y, _ := strconv.Atoi(parts[1])
	return Point{x, y}
}

func simulateSand(blocked map[Point]bool, maxY int, floor bool) (Point, bool) {
	x, y := 500, 0

	for {
		// Check if sand has fallen below all rocks (into abyss)
		if !floor && y > maxY {
			return Point{}, false
		}

		// Try to move down
		if floor && y+1 == maxY+2 {
			// Hit the floor
			return Point{x, y}, true
		} else if !blocked[Point{x, y + 1}] {
			y++
		} else if !blocked[Point{x - 1, y + 1}] {
			// Try to move down-left
			x--
			y++
		} else if !blocked[Point{x + 1, y + 1}] {
			// Try to move down-right
			x++
			y++
		} else {
			// Sand comes to rest
			return Point{x, y}, true
		}
	}
}

func findMaxY(rocks map[Point]bool) int {
	maxY := 0
	for p := range rocks {
		if p.y > maxY {
			maxY = p.y
		}
	}
	return maxY
}

func part1(text string) int {
	rocks := parsePaths(text)
	maxY := findMaxY(rocks)

	// Copy rocks to blocked
	blocked := make(map[Point]bool)
	for p := range rocks {
		blocked[p] = true
	}

	count := 0
	for {
		pos, ok := simulateSand(blocked, maxY, false)
		if !ok {
			break
		}
		blocked[pos] = true
		count++
	}
	return count
}

func part2(text string) int {
	rocks := parsePaths(text)
	maxY := findMaxY(rocks)

	// Copy rocks to blocked
	blocked := make(map[Point]bool)
	for p := range rocks {
		blocked[p] = true
	}

	count := 0
	source := Point{500, 0}
	for {
		pos, _ := simulateSand(blocked, maxY, true)
		blocked[pos] = true
		count++
		if pos == source {
			break
		}
	}
	return count
}

func main() {
	// Get the directory of the executable
	execPath, _ := os.Executable()
	dir := filepath.Dir(execPath)

	// Try to find input.txt in parent directory
	inputPath := filepath.Join(dir, "..", "input.txt")
	if _, err := os.Stat(inputPath); os.IsNotExist(err) {
		// Fallback: try current working directory
		cwd, _ := os.Getwd()
		inputPath = filepath.Join(cwd, "..", "input.txt")
	}

	file, err := os.Open(inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var sb strings.Builder
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		sb.WriteString(scanner.Text())
		sb.WriteString("\n")
	}

	text := sb.String()

	fmt.Println("Part 1:", part1(text))
	fmt.Println("Part 2:", part2(text))
}

File: 2022/day14/rust/solution.rs
use std::collections::HashSet;
use std::fs;
use std::path::Path;

fn parse_paths(text: &str) -> HashSet<(i32, i32)> {
    let mut rocks = HashSet::new();

    for line in text.trim().lines() {
        let points: Vec<(i32, i32)> = line
            .split(" -> ")
            .map(|p| {
                let parts: Vec<i32> = p.split(',').map(|n| n.parse().unwrap()).collect();
                (parts[0], parts[1])
            })
            .collect();

        for i in 0..points.len() - 1 {
            let (x1, y1) = points[i];
            let (x2, y2) = points[i + 1];

            if x1 == x2 {
                // Vertical line
                let (min_y, max_y) = if y1 < y2 { (y1, y2) } else { (y2, y1) };
                for y in min_y..=max_y {
                    rocks.insert((x1, y));
                }
            } else {
                // Horizontal line
                let (min_x, max_x) = if x1 < x2 { (x1, x2) } else { (x2, x1) };
                for x in min_x..=max_x {
                    rocks.insert((x, y1));
                }
            }
        }
    }

    rocks
}

fn simulate_sand(blocked: &HashSet<(i32, i32)>, max_y: i32, floor: bool) -> Option<(i32, i32)> {
    let (mut x, mut y) = (500, 0);

    loop {
        // Check if sand has fallen below all rocks (into abyss)
        if !floor && y > max_y {
            return None;
        }

        // Try to move down
        if floor && y + 1 == max_y + 2 {
            // Hit the floor
            return Some((x, y));
        } else if !blocked.contains(&(x, y + 1)) {
            y += 1;
        } else if !blocked.contains(&(x - 1, y + 1)) {
            // Try to move down-left
            x -= 1;
            y += 1;
        } else if !blocked.contains(&(x + 1, y + 1)) {
            // Try to move down-right
            x += 1;
            y += 1;
        } else {
            // Sand comes to rest
            return Some((x, y));
        }
    }
}

fn part1(text: &str) -> usize {
    let rocks = parse_paths(text);
    let max_y = rocks.iter().map(|(_, y)| *y).max().unwrap();
    let mut blocked = rocks.clone();
    let mut count = 0;

    loop {
        match simulate_sand(&blocked, max_y, false) {
            None => break,
            Some(pos) => {
                blocked.insert(pos);
                count += 1;
            }
        }
    }

    count
}

fn part2(text: &str) -> usize {
    let rocks = parse_paths(text);
    let max_y = rocks.iter().map(|(_, y)| *y).max().unwrap();
    let mut blocked = rocks.clone();
    let mut count = 0;

    loop {
        let pos = simulate_sand(&blocked, max_y, true).unwrap();
        blocked.insert(pos);
        count += 1;
        if pos == (500, 0) {
            break;
        }
    }

    count
}

fn main() {
    let exe_path = std::env::current_exe().unwrap();
    let dir = exe_path.parent().unwrap();
    let input_path = dir.join("..").join("input.txt");

    let text = fs::read_to_string(&input_path)
        .unwrap_or_else(|_| {
            // Fallback: try relative to current directory
            let alt_path = Path::new("../input.txt");
            fs::read_to_string(alt_path).expect("Could not read input.txt")
        });

    println!("Part 1: {}", part1(&text));
    println!("Part 2: {}", part2(&text));
}
